### ISSUE TITLE
Remove curl dependency from AssetsManager, AssetsManagerEx, Downloader

### ISSUE THREAD

**Issue by Yehsam23** [cocos2d/cocos2d-x#10382]
Cocos2d-x v3.4 had a big improved feature.

Smaller Size
Libcurl has been replaced by system network library on mobile platform such as IOS and android, which make game package more lightweight.

But the AssetsManager & Downloader still use curl libs.

Would you change those curl to use HttpClient?

Thanks

**Comment by minggo** [comment]
Yep, we will replace curl in AssetsManager too.
Because it needs more time, so we don't implement it in v3.4.

**Comment by Yehsam23** [comment]
Thanks a lot

**Comment by stammen** [comment]
The Windows 8.1 Universal App version of cocos2d-x would also like curl removed from AssetsManager. We have implemented WinRT native networking but will still need curl for the AssetsManager.

**Comment by walzer** [comment]
:+1: 
Schedule in v3.7

**Comment by ricardoquesada** [comment]
I'll start working on this. I'll implement the "cocos net" abstraction

**Comment by ricardoquesada** [comment]
Native replacements for **iOS**:
- AFNetworking: https://github.com/AFNetworking/AFNetworking
- NSURLConnection / NSURLSession: https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/URLLoadingSystem/URLLoadingSystem.html

Native replacements for **Android**:
- Volley: https://www.youtube.com/watch/?v=yhv8l9F44qo
- OkHTTP: http://square.github.io/okhttp/

Any other ?
@minggo @pandamicro what do you recommend to use? thanks.

**Comment by ricardoquesada** [comment]
@linshun @pandamicro Hey, I need some help. I noticed that we already have some networking code in cocos/network: https://github.com/cocos2d/cocos2d-x/tree/v3/cocos/network

Should AssetManager / DownloadManager use the cocos/network code ( `HttpClient`) ? And I think that was the suggestion made by @Yehsam23. 

If so, I wouldn't need to create the "network" layer... perhaps I only need to create the missing functionality.

Thoughts?

**Comment by ricardoquesada** [comment]
And what is the difference between `AssetsManager` and `AssetsManagerEx` ?
And any particular reason why `AssetsManager` is a subclass of `Node` ?
Thanks.

**Comment by pandamicro** [comment]
## Third party libraries

For native replacement on Android, Cocos Play & Runtime are using : 

android-async-http: https://github.com/loopj/android-async-http

## About HttpClient

> Should AssetManager / DownloadManager use the cocos/network code ( HttpClient) ? And I think that was the suggestion made by @Yehsam23. If so, I wouldn't need to create the "network" layer... perhaps I only need to create the missing functionality.

Yes, that's what @Yehsam23 suggested.

There are couple of problems about HttpClient: 
- It only support single file downloading
- It only store results in the buffer
- We need to maintain the iOS/Android implementation

What we still need:
- Download file into local file
- Asynchronous downloading process
- Downloading event, notifications and error handling
- Simultaneously downloading files (multi thread support)
- Download resuming (from previous unfinished download)

The disadvantage which I see of creating the missing functionality ourselves is high maintain cost.

If we use a mature third party library, we can save a lot of time and our network layer will be more stable. In this case, we should replace the current implementation of HttpClient with the network layer.

## About AssetsManager

> And what is the difference between AssetsManager and AssetsManagerEx ? And any particular reason why AssetsManager is a subclass of Node ?

AssetsManager was a previous implementation, it can download or update a zip package.
[AssetsManagerEx](http://cocos2d-x.org/docs/manual/framework/html5/v3/assets-manager/en) was totally refactored and provide a smart update logic by comparing local and remote differences (file level) and more functionalities like download resuming, simultaneous download, event notification etc.

I'm not sure why AssetsManager is a subclass of Node, here @minggo 

## About libcurl

@yinjimmy proposed that we can reduce its package size by eliminate some functionalities and keep only http/websocket. I'm not very familiar with libcurl, if it's possible and the size can be effectively reduced, it worth try.

For reference, android-async-http's size is about 50kb, while the current libcurl take almost 1mb.

**Comment by ricardoquesada** [comment]
@pandamicro thanks for the detailed info.

So, apparently `HttpClient` is using [libwebsockets](https://github.com/warmcat/libwebsockets) and I guess libwebsockets has all the functionality that we need... (just a guess since libwebsockets is being used by chrome as well)

So, perhaps what I should do is to wrap the missing functionality of libwebsockets into the namespace `cocos::network`.

What do you think?

**Comment by pandamicro** [comment]
Websocket is long connection network library while Http is short connection, so I don't think we can use libwebsockets to implement Http, correct me if I'm wrong, maybe libwebsockets have http capacity.

HttpClient on iOS is using `network/HttpAsynConnection.h`

https://github.com/cocos2d/cocos2d-x/blob/v3/cocos/network/HttpClient-apple.mm

It's using `java.net.HttpURLConnection` on Android

https://github.com/cocos2d/cocos2d-x/blob/v3/cocos/platform/android/java/src/org/cocos2dx/lib/Cocos2dxHttpURLConnection.java

**Comment by ricardoquesada** [comment]
@pandamicro 
you are right!

So, we have:
- `HTTPClient`: that uses native implementations: Does it work on Windows or Linux ? What does it use for those ?
- `WebSocket`: which is a wrapper of libwebsockets. I guess it is available for all our platforms
- `SocketIO`:  which I don't know what it does.

At this point I'm rather confused :(

What is your suggestion? I don't want to create yet another... it seems that we have too many at this point

**Comment by stammen** [comment]
How about Casablanca, a C++ REST API from Microsoft? It works on iOS, Android, Windows, Windows Phone and Linux.

https://casablanca.codeplex.com/documentation

**Comment by pandamicro** [comment]
All three works for all platforms, they serve for different purposes
1. HTTPClient
   
   It's for HTTP connections (short connection), on Windows/Linux it uses libcurl
   It's used by Lua/JS Http connections, it may be used by C++ users
   Short connection is often used for simple message transitions, downloading files, etc
2. WebSocket
   
   It's for WebSocket connections (long connection), it's originally designed for and essentially used by JS bindings. Originally web standard. I doubt any C++ user use this.
   Long connection is often used for MMO games which requires a durable connection for transmit messages all along with game activities
3. Socket IO
   
   It's a wrapper for WebSocket, originally web implementation built up on the WebSocket standard APIs. We have this essentially for JSB needs, because we have it in our web engine. I doubt any C++ user use this.
   
   http://socket.io/

In conclusion, there are two type of network connections: HTTP and Socket.

I think what we need to do about the network layer, is first of all build up the HTTP APIs for a general usage. Then we can think about support socket ( long connections ) in the network layer. Because HTTP connection is much more popular in games than socket connection.

What do you think ?

**Comment by ricardoquesada** [comment]
@stammen thanks for the suggestion. I'll take a look at it.

@pandamicro 
Thanks!

Although now I understand the purpose of the 3 libraries  (I'm still trying to figure out the purpose of SocketIO) I think that our network stack is confusing. In theory no explanation should be needed.

Something like the following would be much better:

``` c++
class NetworkInterface
{
    virtual openWebSocket(...) = 0;
    virtual openHTTPConnection(...) = 0;
    virtual openSocketIO(...) = 0;
}
```

But I don't want to refactor anything at this moment... perhaps for v4.0. But for v3.x, the less changes that we make, the better.

My initial thoughts are that:
- we should not add any new library (unless the current ones don't support the needed features) for DownloadManager and AssetManager
- our code should not use any network library directly. Instead it should use our abstraction interface. So downloadmanager / AssetManager should use our network interface

I'll continue to do research tomorrow. I'll keep posting my thoughts here.

**Comment by pandamicro** [comment]
Thanks ! Riq

**Comment by ricardoquesada** [comment]
So, this is what I'm doing:

**step 1)**
- add a `URLDownload` class to the `cocos2d::network` network. Basically it will be a super simple wrap of libcurl.
- `Downloader` and `AssetManager` will use this new API instead of calling libcurl directly.

**step 2)**
- Add native support for iOS and Android for `URLDownload`

I'm adding my code:  https://github.com/ricardoquesada/cocos2d-x/tree/cocos_net

**Comment by ricardoquesada** [comment]
@pandamicro 
I have a question regarding `Downloader::batchDownloadSync()`.

It only checks to see if the 1st URL of the group supports resume. What if the second URL doesn't support it ?

Or does `batchDownloadSync()` only work with URL that belong to the same server ?
thanks.

**Comment by ricardoquesada** [comment]
@pandamicro @minggo 
One more question... any objection in moving `Downloader` from the `cocos2d::extensions` namespace to the `cocos2d::network` namespace ?

**Comment by pandamicro** [comment]
> I have a question regarding Downloader::batchDownloadSync().
> It only checks to see if the 1st URL of the group supports resume. What if the second URL doesn't support it ?
> Or does batchDownloadSync() only work with URL that belong to the same server ?

Oh, that's an issue, when I designed Downloader, it's only used by AssetsManagerEx. In AssetsManagerEx, batch download can only download files from the same server. So it never cause any issue. But you are right, it can be used to download files from different servers, and we should check download resuming support for the resuming file, not the first file.

> One more question... any objection in moving Downloader from the cocos2d::extensions namespace to the cocos2d::network namespace ?

Downloader isn't used anywhere else except for AssetsManagerEx, so I think it's safe to refactor its APIs, change its namespace.

**Comment by ricardoquesada** [comment]
ok. thanks for the info. I'll move the `Downloader` then. Thanks.

**Comment by liuzhi1989gbc** [comment]
## 需求

游戏对网络的需求有以下几种：

-   和服务端交互
    -   通过 HTTP
    -   通过 Raw Socket
    -   通过 WebSocket
-   从服务器获取资源

特别说明：

-   Raw Socket 是指客户端和服务端创建 Socket 连接后，双方自行处理数据包的拆包、合并等操作。
-   WebSocket 在握手阶段通过 HTTP 协议进行；握手成功后，和客户端创建 Socket 连接。在通讯时，WebSocket 协议已经处理了拆包、合并等工作。
-   SocketIO 仅仅只是 WebSocket 的一个逻辑层面封装，提供更容易使用 API。所以 SocketIO 仍然属于 WebSocket，而不是 Raw Socket。
-   开发商选择 Raw Socket 还是 WebSocket，很大程度上是由服务端架构决定的。

所以 cocos 实际上只提供了 HTTP 和 WebSocket 两种网络通讯方式的支持，并不支持 Raw Socket。

<br />

## API 设计

用户对三种网络通讯方式的使用，是一种组合的方式，通常会用到其中两种。例如一些新的 MMO 游戏，大多数时候使用 HTTP 通讯，实时 PK 时才使用 RawSocket/WebSocket。

所以在设计 API 时，要避免 All-in-One 的做法。用户不想为自己不需要的功能无谓增加 .so 体积。

其次，每一个 class 应该专注于单一的职责，但需要提供更容易使用的接口。

我认为 network 层的设计应该类似如下：

```cpp

namespace cc.networks
{

// HTTP

enum HttpRequestMethod
{
    get,
    post,
    put,
    delete
}

HttpRequestBasicImpl
{
    void setRequestUrl(string url);
    void setRequestMethod(HttpRequestMethod requestMethod);
    void setTimeout(int timeout);

    void setHeaders(string headers); // set headers string
    void setCookies(string cookies);
    void setPosts(string posts);

    void addHeaderPair(string key, string value); // append pair to headers
    void addCookiePair(string key, string value);
    void addPostPair(string key, string value);

    string getHeadersString();
    string getCookiesString();
    string getPostsString();
}

HttpRequest : HttpRequestBasicImpl
{
    static HttpRequest* createWithUrl(string url);

    void addListener(HttpRequestListener *listener);
    void start();
    void cancel();
}

HttpRequestListener
{
    void onStateChanged(HttpRequest *request, string phase);
    // invoke when every chunk loaded from remote
    void onProgress(HttpRequest *request,
            HttpResponse *response);
    void onCompleted(HttpResponse *response);
}

enum HttpResponseBufferMode
{
    append,
    chunk
};

HttpResponse
{
    static HttpResponse* create(HttpResponseBufferMode mode = append);

    void setBufferMode(HttpResponseBufferMode mode);

    int getCode(); // http response code
    string getString(); // http response string

    int getTotalSize(); // http response total size, maybe -1
    int getLoadedSize();

    int getChunkOffset();
    int getChunkLength();
    unsigned char* getBuffer();
    void flush();
}

HttpRequestQueueBasicImpl
{
    void setBatchSize(int size);
    int getCount();

    void start();
    void cancel();
    void cancelWithTag(int tag);
}

HttpRequestQueue : HttpRequestQueueBasicImpl
{
    static HttpRequestQueue* createWithBatchSize(int size);

    int addRequest(HttpRequest *request); // return tag
}


// WebSocket

WebSocketConnect : HttpRequestBasicImpl
{
    static WebSocketConnect* createWithUrl(string url);

    void addListener(WebSocketConnectListener *listener);
    void open();
    void close();

    string getPhase();

    void sendTextMessage(string message);
    void sendBinaryMessage(unsigned char *buffer, int size);
    void sendMessage(WebSocketConnectMessage *message);
}

WebSocketConnectListener
{
    void onStateChanged(WebSocketConnect *connect, string phase);
    void onSent(WebSocketConnect *connect,
            WebSocketConnectMessage *message);
    void onReceived(WebSocketConnect *connect,
            WebSocketConnectMessage *message);
}

enum WebSocketConnectMessageType
{
    text,
    binary
}

WebSocketConnectMessage
{
    static *WebSocketConnectMessage createWithText(string message);
    static *WebSocketConnectMessage createWithBinary(unsigned char *buffer,
            int size);

    WebSocketConnectMessageType getType();

    string getTextMessage();
    unsigned char *getBinaryMessage();
    int getMessageSize();
}

// Downloader

Downloader : HttpRequestQueueBasicImpl, HttpRequestListener
{
    static HttpRequestQueue* createWithBatchSize(int size);

    // return tag
    int addUrl(string url, string localpath, int offset = 0);

    void flush(); // remove incompleted files
}

```

临时想的设计，大概就这样了。

**Comment by dualface** [comment]
## 需求

游戏对网络的需求有以下几种：
-   和服务端交互
  -   通过 HTTP
  -   通过 Raw Socket
  -   通过 WebSocket
-   从服务器获取资源

特别说明：
-   Raw Socket 是指客户端和服务端创建 Socket 连接后，双方自行处理数据包的拆包、合并等操作。
-   WebSocket 在握手阶段通过 HTTP 协议进行；握手成功后，和客户端创建 Socket 连接。在通讯时，WebSocket 协议已经处理了拆包、合并等工作。
-   SocketIO 仅仅只是 WebSocket 的一个逻辑层面封装，提供更容易使用 API。所以 SocketIO 仍然属于 WebSocket，而不是 Raw Socket。
-   开发商选择 Raw Socket 还是 WebSocket，很大程度上是由服务端架构决定的。

所以 cocos 实际上只提供了 HTTP 和 WebSocket 两种网络通讯方式的支持，并不支持 Raw Socket。

## Needs

Requirements from game for network are the following:
- Interact with the server
  - via HTTP
  - via Raw Socket
  - via WebSocket
- Acquire resources from the server

Special notes：
- Raw Socket occupy only the socket connection, all the manipulation of network package is done manually in both server and client side.
- WebSocket use HTTP protocol to proceed the handshake procedure, once the handshake succeed, it will create a Socket connection between the server and client. During the communication, WebSocket occupy the tasks like de-package, reconstruct, this is the main difference between RawSocket and WebSocket.
- SocketIO is only a logic level wrapper of WebSocket for providing easy to use API. So SocketIO is WebSocket.
- Developers choice between Raw Socket and WebSocket depends mainly on their server structure.

So cocos actually only have HTTP, WebSocket, there is no Raw Socket implementation.

## API 设计

用户对三种网络通讯方式的使用，是一种组合的方式，通常会用到其中两种。例如一些新的 MMO 游戏，大多数时候使用 HTTP 通讯，实时 PK 时才使用 RawSocket/WebSocket。

所以在设计 API 时，要避免 All-in-One 的做法。用户不想为自己不需要的功能无谓增加 .so 体积。

其次，每一个 class 应该专注于单一的职责，但需要提供更容易使用的接口。

我认为 network 层的设计应该类似如下：

## API Design

Developers often use multiple network services in their game, for example, some MMO games use HTTP under most conditions, but they also use RawSocket/WebSocket for real time PK battle.

So when design the API, it's better to avoid an All-in-One way. Developers don't want to increase their package size for something they don't need.

Secondly, every class should focus on single responsibility with easy to use API.

So I think network layer API design could be like the following: 

``` cpp

namespace cc.network

// HTTP

enum HttpRequestMethod
{
    get,
    post,
    put,
    delete
}

HttpRequestBasicImpl
{
    void setRequestUrl(string url);
    void setRequestMethod(HttpRequestMethod requestMethod);
    void setTimeout(int timeout);

    void setHeaders(string headers); // set headers string
    void setCookies(string cookies);
    void setPosts(string posts);

    void addHeaderPair(string key, string value); // append pair to headers
    void addCookiePair(string key, string value);
    void addPostPair(string key, string value);

    string getHeadersString();
    string getCookiesString();
    string getPostsString();
}

HttpRequest : HttpRequestBasicImpl
{
    static HttpRequest* createWithUrl(string url);

    void addListener(HttpRequestListener *listener);
    void start();
    void cancel();
}

HttpRequestListener
{
    void onStateChanged(HttpRequest *request, string phase);
    // invoke when every chunk loaded from remote
    void onProgress(HttpRequest *request,
            HttpResponse *response);
    void onCompleted(HttpResponse *response);
}

enum HttpResponseBufferMode
{
    append,
    chunk
};

HttpResponse
{
    static HttpResponse* create(HttpResponseBufferMode mode = append);

    void setBufferMode(HttpResponseBufferMode mode);

    int getCode(); // http response code
    string getString(); // http response string

    int getTotalSize(); // http response total size, maybe -1
    int getLoadedSize();

    int getChunkOffset();
    int getChunkLength();
    unsigned char* getBuffer();
    void flush();
}

HttpRequestQueueBasicImpl
{
    void setBatchSize(int size);
    int getCount();

    void start();
    void cancel();
    void cancelWithTag(int tag);
}

HttpRequestQueue : HttpRequestQueueBasicImpl
{
    static HttpRequestQueue* createWithBatchSize(int size);

    int addRequest(HttpRequest *request); // return tag
}


// WebSocket

WebSocketConnect : HttpRequestBasicImpl
{
    static WebSocketConnect* createWithUrl(string url);

    void addListener(WebSocketConnectListener *listener);
    void open();
    void close();

    string getPhase();

    void sendTextMessage(string message);
    void sendBinaryMessage(unsigned char *buffer, int size);
    void sendMessage(WebSocketConnectMessage *message);
}

WebSocketConnectListener
{
    void onStateChanged(WebSocketConnect *connect, string phase);
    void onSent(WebSocketConnect *connect,
            WebSocketConnectMessage *message);
    void onReceived(WebSocketConnect *connect,
            WebSocketConnectMessage *message);
}

enum WebSocketConnectMessageType
{
    text,
    binary
}

WebSocketConnectMessage
{
    static *WebSocketConnectMessage createWithText(string message);
    static *WebSocketConnectMessage createWithBinary(unsigned char *buffer,
            int size);

    WebSocketConnectMessageType getType();

    string getTextMessage();
    unsigned char *getBinaryMessage();
    int getMessageSize();
}

// Downloader

Downloader : HttpRequestQueueBasicImpl, HttpRequestListener
{
    static HttpRequestQueue* createWithBatchSize(int size);

    // return tag
    int addUrl(string url, string localpath, int offset = 0);

    void flush(); // remove incompleted files
}

```

临时想的设计，大概就这样了。

**Comment by pandamicro** [comment]
@ricardoquesada I translated @dualface 's proposition. Please take a look

**Comment by ricardoquesada** [comment]
@pandamicro @dualface 

Thanks. I like it. 

My comments are:
Currently we already have an `HTTPClient` that has an `HTTPRequest` class.
And we also have a complete implementation of `WebSocket`, which is not a subclass `HTTPRequest....`.

I think your proposal is very good, but it is outside the scope of this feature since it will require changes that might break compatibility.

For v3.8, I would just remove libcurl dependency, with minimal changes.
For v4.0, I would reorganize the network code.

thoughts?

**Comment by pandamicro** [comment]
I agree, we can minify the impacts in v3.8 and refactor it well in v4.0

**Comment by ricardoquesada** [comment]
Ok. I'm still working on this feature. I got somewhat delayed because I was trying to understand the whole code... I think more or less I know how it works. 
Isolating the libcurl code will take me one or two days, and then replacing libcurl with native libraries will take about 5 days (although next week I'll be in a conference).

(my code is here: https://github.com/ricardoquesada/cocos2d-x/tree/cocos_net)

**Comment by ricardoquesada** [comment]
I also added a few optimizations in the code.
What's next:

libcurl isolation:
- detect if the code is already running in the cocos thread (for sync operations). If so no need to call `performInCocosThread()`
- make it work... Need to fix a few bugs int he callbacks

replace libcurl with native:
- research: how to use the native networking libraries
- replace libcurl on Android and iOS

**Comment by pandamicro** [comment]
Great, as the release date of v3.8 is synced with Cocos v2.3.2 at about 08.26

I have scheduled Cocos2d-x v3.8 releases as the following:
- 8.10 v3.8 Beta
- 8.19 v3.8 RC0
- 8.26 v3.8 Final

We already submit a version to be tested by QA team.
Beta and RC version will be released in the community.

**Comment by ricardoquesada** [comment]
ok. thanks for the info.

**Comment by ricardoquesada** [comment]
@pandamicro @minggo 

So, I think I finished the first part. And it seems that it is working... but I need to test more.
Where are the tests cases for `Downloader` and `AssetManagerEx` ?

I could only find one, in: cpp-tests->extensions->AssetManagerExTest

But that test only tests `Downloader::downloadAsync()` and I need to also test:
- `downloadToBufferAsync()`
- `downloadToBufferSync()`
- `downloadSync()`
- `batchDownloadAsync()`
- `batchDownloadSync()`

Thanks.

**Comment by pandamicro** [comment]
- `downloadToBufferAsync` and `downloadToBuffer` can be tested with JSB's `jsb.loadRemoteImg(url, function (succeed, tex) {})`. The test case is in js-tests/src/TextureCacheTest/TextureCacheTest.js
- `batchDownloadAsync` and `batchDownload` can be tested with AssetsManagerExTest, because it will download simultaneously all the files to be updated.

As for `downloadSync`, `batchDownloadSync`, `downloadToBufferSync`, I didn't wrote a test case, because they simply use correspond download functions.

**Comment by minggo** [comment]
@ricardoquesada 
Sorry, i am not familiar with this part of code. It is designed and implemented by @pandamicro .

**Comment by ricardoquesada** [comment]
@pandamicro ok. thanks for the info.

**Comment by ricardoquesada** [comment]
@pandamicro 
First part finished: https://github.com/cocos2d/cocos2d-x/pull/13281

I'll update the linux, android and windows projects on Monday.
Please, let me know if you like the changes that I did. Thanks!

**Comment by pandamicro** [comment]
@yangws will continue the second part work in v3.9

**Comment by ricardoquesada** [comment]
Thanks.

@yangws Things that I would do:
- Rename `CCDownloaderImpl.*` to `CCDownloaderImplCurl.*`
- Create a `CCDownloaderImplAndroid.*` with the android code
- Create a `CCDownloaderImplIOS.*` with the iOS code

and I guess that's it.

**Comment by felixjingga** [comment]
Please see #13672 I think there are some issues after the rewrite in v3.8.. If the manifests contains hundred of files, it "seek" the files and make all the process slower and looks "not responding" to user. Another one was if there's no internet connection, on some devices it will crash. I'm trying this on Android.. so maybe @ricardoquesada @yangws can take a look at this issue.

Also, is it possible to improve the assetmanager so that it will send the current version of its assets, or maybe the ability to call a callback before it send the request? The use case can be like this:
Assume we are simulating a CDN, we need the assetmanager can request to the server not to one base URL, but to different base url (somekind like serverX.url.com, where X is 1 ... N) or another use case is to set a header-data.

**Comment by pandamicro** [comment]
@felixjingga Thanks for reporting this, the issue #13672 is scheduled to v3.9

**Comment by pandamicro** [comment]
Fixed by https://github.com/cocos2d/cocos2d-x/pull/13832
